test/ecto/query/planner_test.exs

```diff
@@ -54,12 +60,18 @@ defmodule Ecto.Query.PlannerTest do
       field :posted, :naive_datetime
       field :visits, :integer
       field :links, {:array, Custom.Permalink}
+      field :crazy_post, :string
       has_many :comments, Ecto.Query.PlannerTest.Comment
       has_many :extra_comments, Ecto.Query.PlannerTest.Comment
       has_many :special_comments, Ecto.Query.PlannerTest.Comment, where: {Ecto.Query.PlannerTest.Comment, :special, []}
+      many_to_many :crazy_comments, Comment, join_through: CommentPost, where: {Comment, :crazy, []}
 
       many_to_many :shared_special_comments, Comment, join_through: CommentPost, where: {Comment, :special, []}, join_through_where: {CommentPost, :inactive, []}
     end
+
+    def crazy() do
+      Ecto.Query.dynamic([row], row.crazy_post == "crazy")
```

Comment:
Have you tried a test that does not use a literal? Something like:

```elixir
many_to_many ..., where: {Comment, :crazy, ["crazy"]}

def crazy(value) do
 Ecto.Query.dynamic([row], row.crazy_post == ^value)
end
```

I want to make sure we handle params correctly.

Reply:
I'll add a case for that in the morning :)

Reply:
So that actually doesn't handle params correctly, but also requires a bit more thought. I can't just protected against `{:^, _, ix}` and rewrite that, because the prewalk could find that for other reasons, right? It would seem to me like I'd need to protect against

Reply:
Actually, I think I'm looking at the wrong pattern there, so nevermind.

Reply:
Okay, so I was originally hesitant to guard against `{:^, _, ix}` because I was worried of what would happen if someone literally interpolated `^0`, but I'm realizing now that isn't how that would work. It would instead be set as a parameter in this same way.

Reply:
@michalmuskala got it. I had to additionally rewrite parameters. I imagine there may be more rewrites that need to happen here, but I'm not sure what they are. Those fixes *could* be done in separate PRs as well, if there are any. I'm happy to do them in this one as well though.